File: 18-metaprogramming.py
class Settings:
    def __init__(self, values):
        self.values = values

    def __getattr__(self, name):
        if name in self.values:
            return self.values[name]

        raise AttributeError(f"Unknown setting: {name}")


settings = Settings({
    "host": "localhost",
    "port": 3000,
})

print(settings.host)
print(settings.port)

print("---")


def add_predicate(class_object, field_name):
    def predicate(self):
        return bool(getattr(self, field_name))

    setattr(class_object, f"has_{field_name}", predicate)


class User:
    def __init__(self, name, email):
        self.name = name
        self.email = email


add_predicate(User, "email")

user = User("Ada", "ada@example.com")

print(user.has_email())

File: 18-metaprogramming.rb
# Ruby metaprogramming 可以喺 runtime 攔截 method 或動態加 method。
class Settings
  def initialize(values)
    @values = values
  end

  def method_missing(name, *)
    return @values[name] if @values.key?(name)

    super
  end

  def respond_to_missing?(name, include_private = false)
    @values.key?(name) || super
  end
end

settings = Settings.new({
                          host: 'localhost',
                          port: 3000
                        })

puts settings.host
puts settings.port

puts '---'

# define_method 可以根據欄位名即場幫 class 加 predicate method。
def add_predicate(class_object, field_name)
  class_object.define_method("has_#{field_name}?") do
    !!public_send(field_name)
  end
end

class User
  attr_reader :name, :email

  def initialize(name, email)
    @name = name
    @email = email
  end
end

add_predicate(User, :email)

user = User.new('Ada', 'ada@example.com')

puts user.has_email?
